### PR TITLE
[NN] Add EGT Layer

### PR DESCRIPTION
**Pull request by ZHITENGLI**
## Description

We'd like to add EGTLayer to the graph transformer family in DGL, which is the basic unit of Edge-augmented Graph Transformer (EGT).

## Checklist
Please feel free to remove inapplicable items for your PR.
- [x] The PR title starts with [$CATEGORY] (such as [NN], [Model], [Doc], [Feature]])
- [x] I've leverage the [tools](https://docs.google.com/document/d/1iHyj7zlmygKSk5gBPsqIqL5ASPzJSPREaNT_QdsiYA4/edit) to beautify the python and c++ code.
- [ ] The PR is complete and small, read the [Google eng practice (CL equals to PR)](https://google.github.io/eng-practices/review/developer/small-cls.html) to understand more about small PR. In DGL, we consider PRs with less than 200 lines of core code change are small (example, test and documentation could be exempted).
- [x] All changes have test coverage
- [x] Code is well-documented
- [x] To the best of my knowledge, examples are either not affected by this change, or have been fixed to be compatible with this change

### CODE REVIEW

**Comment by rudongyu** [comment]
@dgl-bot

File: python/dgl/nn/pytorch/gt/egt.py [hunk 1]
```diff
@@ -0,0 +1,178 @@
...
+
+    Parameters
+    ----------
+    ndim : int
+        Node embedding dimension.
+    edim : int
+        Edge embedding dimension.
+    num_heads : int
+        Number of attention heads, by which :attr: `ndim` is divisible.
+    num_vns : int
```
Comment (rudongyu) [inline python/dgl/nn/pytorch/gt/egt.py:21]: directly use "num_virtual_nodes" for clarity?

File: python/dgl/nn/pytorch/gt/egt.py [hunk 2]
```diff
@@ -0,0 +1,178 @@
...
+        Number of virtual nodes.
+    dropout : float, optional
+        Dropout probability. Default: 0.0.
+    attn_dropout : float, optional
+        Attention dropout probability. Default: 0.0.
+    activation : callable activation layer, optional
+        Activation function. Default: nn.ELU().
+    ffn_multiplier : float, optional
```
Comment (rudongyu) [inline python/dgl/nn/pytorch/gt/egt.py:29]: ffn_hidden_size_multiplier? Shall we elaborate more on how it affects the hidden size of feed forward layers for node and edge update?

File: python/dgl/nn/pytorch/gt/egt.py [hunk 3]
```diff
@@ -0,0 +1,178 @@
...
+        activation=nn.ELU(),
+        ffn_multiplier=2.0,
+        edge_update=True,
+    ):
+        super().__init__()
+        self.num_heads = num_heads
+        self.num_vns = num_vns
+        self.edge_update = edge_update
+
+        assert not (ndim % num_heads)
```
Comment (rudongyu) [inline python/dgl/nn/pytorch/gt/egt.py:71]: add a description for assertion error

File: python/dgl/nn/pytorch/gt/egt.py [hunk 4]
```diff
@@ -0,0 +1,178 @@
...
+                nn.Linear(edim, edge_inner_dim),
+                activation,
+                nn.Linear(edge_inner_dim, edim),
+                nn.Dropout(dropout),
+            )
+
+    def forward(self, h, e, mask=None):
+        """Forward computation. Note: :attr:`h` and :attr:`e` should be padded
+        with embedding of virtual nodes if :attr:`num_vns` > 0, while
+        :attr:`mask` should be padded with `0` values for virtual nodes.
```
Comment (rudongyu) [inline python/dgl/nn/pytorch/gt/egt.py:106]: the padding should be in the beginning or at the end?

File: python/dgl/nn/pytorch/gt/egt.py [hunk 5]
```diff
@@ -0,0 +1,178 @@
...
+
+        Parameters
+        ----------
+        h : torch.Tensor
+            A 3D input tensor. Shape: (batch_size, N, :attr:`ndim`), where N
+            is the sum of maximum number of nodes and number of virtual nodes.
```
Comment (rudongyu) [inline python/dgl/nn/pytorch/gt/egt.py:112]:
```suggestion
            is the sum of the maximum number of nodes and the number of virtual nodes.
```

File: python/dgl/nn/pytorch/gt/egt.py [hunk 6]
```diff
@@ -0,0 +1,178 @@
...
+
+        h_r1 = h
+        e_r1 = e
+
+        h_ln = self.mha_ln_h(h)
+        e_ln = self.mha_ln_e(e)
+        QKV = self.QKV(h_ln)
+        E = self.E(e_ln)
+        G = self.G(e_ln)
+        shp = QKV.shape
```
Comment (rudongyu) [inline python/dgl/nn/pytorch/gt/egt.py:138]: I'd recommend to use meaningful names. e.g.,
```
bsz, N, _ = QKV.shape
...
```

File: python/dgl/nn/pytorch/gt/egt.py [hunk 7]
```diff
@@ -0,0 +1,178 @@
...
+
+        assert not (ndim % num_heads)
+        self.dot_dim = ndim // num_heads
+        self.mha_ln_h = nn.LayerNorm(ndim)
+        self.mha_ln_e = nn.LayerNorm(edim)
+        self.E = nn.Linear(edim, num_heads)
+        self.QKV = nn.Linear(ndim, ndim * 3)
+        self.G = nn.Linear(edim, num_heads)
+        self.attn_dropout = nn.Dropout(attn_dropout)
+        self.O_h = nn.Linear(ndim, ndim)
```
Comment (rudongyu) [inline python/dgl/nn/pytorch/gt/egt.py:79]: Could we use module names like "edge_input_linear, node_input_linear, gate_linear, etc." for readability?

**Review by rudongyu — CHANGES_REQUESTED** [review]
I've done a pass.

File: python/dgl/nn/pytorch/gt/egt.py (hunk 2 above)
Comment (ZHITENGLI) [reply python/dgl/nn/pytorch/gt/egt.py:29]: Since `ffn_multiplier` is set to 1.0 for all configs, I'd like to remove it later.

File: python/dgl/nn/pytorch/gt/egt.py (hunk 4 above)
Comment (ZHITENGLI) [reply python/dgl/nn/pytorch/gt/egt.py:106]: Actually it doesn't matter since users can define which nodes to be used in the readout function in EGT model. The official implementation of EGT is to put it at the beginning.

File: python/dgl/nn/pytorch/gt/egt.py (hunk 2 above)
Comment (rudongyu) [reply python/dgl/nn/pytorch/gt/egt.py]: ok, fair enough.

File: python/dgl/nn/pytorch/gt/egt.py (hunk 4 above)
Comment (rudongyu) [reply python/dgl/nn/pytorch/gt/egt.py]: It seems that the `degree_scalars` of virtual nodes is set to 1.0 at the beginning. Does it suggest that padding at the beginning is mandatory?

Comment (rudongyu) [reply python/dgl/nn/pytorch/gt/egt.py]: btw, do you think it's a better solution to include a VirtualNode module like what EGT does? Will it reduce user's burden and improve usability?

Comment (ZHITENGLI) [reply python/dgl/nn/pytorch/gt/egt.py]: > It seems that the `degree_scalars` of virtual nodes is set to 1.0 at the beginning. Does it suggest that padding at the beginning is mandatory?

Yes, I was wrong. It should be in the beginning.

Comment (ZHITENGLI) [reply python/dgl/nn/pytorch/gt/egt.py]: > btw, do you think it's a better solution to include a VirtualNode module like what EGT does? Will it reduce user's burden and improve usability?

I think so, but this VirtualNode module may not be universal since its function is not only to generate virtual node embedding, but also to general positional encodings or attention bias related to the virtual nodes. While PEs and attention bias are different in different models.

File: python/dgl/nn/pytorch/gt/egt.py [hunk 8]
```diff
@@ -0,0 +1,178 @@
...
+            invalid positions are indicated by `-inf`.
+            Shape: (batch_size, N, N). Default: None.
+
+        Returns
+        -------
+        h : torch.Tensor
+            The output node embedding. Shape: (batch_size, N, :attr:`feat_size`).
+        e : torch.Tensor, optional
+            The output edge embedding. Shape: (batch_size, N, N, :attr:`edge_feat_size`).
+            It is only returned when :attr:`edge_update` is True.
```
Comment (rudongyu) [inline python/dgl/nn/pytorch/gt/egt.py:125]:
```suggestion
            It is returned only if :attr:`edge_update` is True.
```

File: python/dgl/nn/pytorch/gt/egt.py [hunk 9]
```diff
@@ -0,0 +1,178 @@
...
+        """
+
```
Comment (rudongyu) [inline python/dgl/nn/pytorch/gt/egt.py:127]:
```suggestion

```

**Review by rudongyu — COMMENTED** [review]
some minors

**Comment by rudongyu** [comment]
@dgl-bot

**Comment by rudongyu** [comment]
@dgl-bot

**Comment by rudongyu** [comment]
@dgl-bot

**Review by rudongyu — APPROVED** [review]
Thanks for your great effort!
